eva-accession-import-automation/genome_downloader.py

```diff
@@ -0,0 +1,170 @@
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from get_assembly_report_url import *
+import os
+import pandas
+import sys
+import urllib.request
+import re
+import json
+from __init__ import *
+
+
+def get_header_line_index(assembly_report_path):
+    try:
+        with open(assembly_report_path) as assembly_report_file_handle:
+            return next(line_index for line_index, line in enumerate(assembly_report_file_handle)
+                        if line.lower().startswith("# sequence-name") and "sequence-role" in line.lower())
+    except StopIteration:
+        raise Exception("Could not locate header row in the assembly report!")
+
+
+def get_dataframe_for_assembly_report(assembly_report_path):
+    return pandas.read_csv(assembly_report_path, skiprows=get_header_line_index(assembly_report_path),
+                           dtype=str, sep='\t')
+
+
+def download_assembly_report(assembly_accession, directory_path):
+    assembly_report_url = get_assembly_report_url(assembly_accession)
+    assembly_report_path = directory_path + os.path.basename(assembly_report_url)
+    os.makedirs(os.path.dirname(assembly_report_path), exist_ok=True)
+    urllib.request.urlretrieve(assembly_report_url, assembly_report_path)
+    urllib.request.urlcleanup()
+    return assembly_report_path
+
+
+def get_sequence_from_ncbi(accession, fasta_path, eutils_api_key, written_contigs):
+    url = 'https://eutils.ncbi.nlm.nih.gov/entrez/eutils/efetch.fcgi?db=nuccore&id=' + accession + \
+           '&rettype=fasta&retmode=text&api_key=' + eutils_api_key + '&tool=eva&email=eva-dev@ebi.ac.uk'
+    logger.info('Downloading ' + accession)
+    with urllib.request.urlopen(url) as response:
+        data = get_and_process_sequence(response)
+        if is_sequence_valid(data):
+            logger.info('FASTA sequence not available for' + accession)
+        else:
+            contatenate_sequence_to_fasta(fasta_path, data)
+            logger.info(accession + " downloaded and added to FASTA file")
+            written_contigs.append(accession)
+
+
+def get_and_process_sequence(response):
+    retrieved_data = response.read().decode('utf-8')
+    # Delete accession prefixes
+    processed_data = re.sub(r'\s*ENA\|.*\|', '', retrieved_data)
+    return processed_data
+
+
+def is_sequence_valid(data):
+    # If a file has more than one line, then it is concatenated into the full assembly FASTA file
+    # (empty sequences can't be indexed)
+    is_sequence_empty = re.match(r'^>.*\n[ACGTN]+', data) is None
```

Comment:
I am not sure what the purpose of this is... Is this to avoid complications later when the FASTA file is used by eva-pipeline or other pipelines?

Reply:
That was a way to check the sequence is not empty so we don't add empty sequences to the FASTA file which will give problems in later stages. But this function was removed and empty validation moved to `get_sequence_from_ncbi` function